xl2times/transforms.py

```diff
@@ -501,81 +498,67 @@ def process_flexible_import_table(
         data_columns = [x for x in df.columns if x not in known_columns]
 
-        # TODO: Replace this with something similar to know columns from config
-        # Populate index columns
-        index_columns = [
-            "region",
-            "process",
-            "commodity",
-            "commodity-in",
-            "commodity-in-aux",
-            "commodity-out",
-            "commodity-out-aux",
-            "attribute",
-            "year",
-            "timeslice",
-            "limtype",
-            "currency",
-            "other_indexes",
-        ]
+        # Populate index columns (same as known columns for this table type)
+        index_columns = known_columns
         for colname in index_columns:
             if colname not in df.columns:
-                df[colname] = [None] * nrows
+                df[colname] = None
         table = replace(table, dataframe=df)
 
         df = table.dataframe
 
-        attribute = "attribute"
-        df, attribute_suffix = utils.explode(df, data_columns)
-
-        # Append the data column name to the Attribute column values
-        if nrows > 0:
-            i = df[attribute].notna()
-            df.loc[i, attribute] = df.loc[i, attribute] + "~" + attribute_suffix[i]
-            i = df[attribute].isna()
-            df.loc[i, attribute] = attribute_suffix[i]
+        if data_columns:
+            df, attribute_suffix = utils.explode(df, data_columns)
+            # Append the data column name to the Attribute column values
+            i = df["attribute"].notna()
+            df.loc[i, "attribute"] = df.loc[i, "attribute"] + "~" + attribute_suffix[i]
+            i = df["attribute"].isna()
+            df.loc[i, "attribute"] = attribute_suffix[i]
 
         # Capitalise all attributes, unless column type float
-        if df[attribute].dtype != float:
-            df[attribute] = df[attribute].str.upper()
+        if df["attribute"].dtype != float:
+            df["attribute"] = df["attribute"].str.upper()
 
         # Handle Attribute containing tilde, such as 'STOCK~2030'
-        for attr in df[attribute].unique():
-            if "~" in attr:
-                i = df[attribute] == attr
+        index = df["attribute"].str.contains("~")
+        if any(index):
+            for attr in df["attribute"][index].unique():
+                i = index & (df["attribute"] == attr)
                 parts = attr.split("~")
                 for value in parts:
                     colname, typed_value = _get_colname(value, legal_values)
                     if colname is None:
-                        df.loc[i, attribute] = typed_value
+                        df.loc[i, "attribute"] = typed_value
                     else:
                         df.loc[i, colname] = typed_value
 
         # Handle Other_Indexes
         other = "other_indexes"
-        for attr in df[attribute].unique():
+        for attr in df["attribute"].unique():
             if attr == "END":
```

Comment:
This loop is redundant since it only does something for the case `attr == "END"`. Might as well replace it with `if "END" in df[attribute]:`

Reply:
Thanks!